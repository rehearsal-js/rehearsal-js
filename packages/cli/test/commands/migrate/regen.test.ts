/* eslint-disable @typescript-eslint/no-explicit-any */
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { createLogger, format, transports } from 'winston';

import {
  initTask,
  depInstallTask,
  convertTask,
  tsConfigTask,
  lintConfigTask,
  regenTask,
} from '../../../src/commands/migrate/tasks/index.js';
import { prepareTmpDir, listrTaskRunner, createMigrateOptions } from '../../test-helpers/index.js';

const logger = createLogger({
  transports: [new transports.Console({ format: format.cli() })],
});

describe('Task: regen', async () => {
  let basePath = '';
  let output = '';
  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });
  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
  });

  beforeEach(() => {
    output = '';
    basePath = prepareTmpDir('basic');
  });

  afterEach(() => {
    output = '';
    vi.clearAllMocks();
  });

  // test('throw error with no tsconfig.json', async () => {
  //   const options = createMigrateOptions(basePath);
  //   const tasks = [await initTask(options), await regenTask(options, logger)];

  //   try {
  //     await listrTaskRunner(tasks);
  //   } catch (error: any) {
  //     expect(error.message).toContain(`Config file 'tsconfig.json' not found`);
  //   }
  // });

  // test('no effect on JS filse before conversion', async () => {
  //   const options = createMigrateOptions(basePath);
  //   const tasks = [
  //     await initTask(options),
  //     await tsConfigTask(options),
  //     await regenTask(options, logger),
  //   ];

  //   await listrTaskRunner(tasks);
  //   expect(output).matchSnapshot();
  // });

  test('update ts and lint errors based on previous conversion', async () => {
    const options = createMigrateOptions(basePath);
    const tasks = [
      await initTask(options),
      await depInstallTask(options),
      await tsConfigTask(options),
      await lintConfigTask(options),
      await convertTask(options, logger),
      await regenTask(options, logger),
    ];

    await listrTaskRunner(tasks);
    console.warn(output);
    expect(output).matchSnapshot();
  });
});
