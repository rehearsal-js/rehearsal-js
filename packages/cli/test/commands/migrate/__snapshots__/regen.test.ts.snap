// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: regen > no effect on JS filse before conversion 1`] = `
"[STARTED] Initialize
[DATA] Setting up config for basic
[SUCCESS] Initialize
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Regenerating report for TS errors and Eslint errors
[TITLE] Migration Report Generated
[TITLE] 
[TITLE] 
[TITLE]   0 ts files have been scanned
[TITLE] 
[TITLE]   0 errors caught by rehearsal
[TITLE] 
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE] 
[TITLE]     -- 0 eslint errors, with details in the report
[TITLE] 
[SUCCESS] Migration Report Generated
[SUCCESS] 
[SUCCESS] 
[SUCCESS]   0 ts files have been scanned
[SUCCESS] 
[SUCCESS]   0 errors caught by rehearsal
[SUCCESS] 
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS] 
[SUCCESS]     -- 0 eslint errors, with details in the report
[SUCCESS] 
"
`;

exports[`Task: regen > update ts and lint errors based on previous conversion 1`] = `
"[STARTED] Initialize
[DATA] Setting up config for basic
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Analyzing Project
[DATA] Running migration on basic
[SUCCESS] Analyzing Project
[STARTED] Convert JS files to TS
[DATA] git mv /foo.js to /foo.ts
[DATA] git mv /depends-on-foo.js to /depends-on-foo.ts
[DATA] git mv /index.js to /index.ts
[DATA] processing file: /foo.ts
[DATA] processing file: /depends-on-foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Complete
[TITLE] 
[TITLE] 
[TITLE]   3 JS files converted to TS
[TITLE] 
[TITLE]   4 errors caught by rehearsal
[TITLE] 
[TITLE]   3 have been fixed by rehearsal
[TITLE] 
[TITLE]   1 errors need to be fixed manually
[TITLE] 
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE] 
[TITLE]     -- 1 eslint errors, with details in the report
[TITLE] 
[SUCCESS] Migration Complete
[SUCCESS] 
[SUCCESS] 
[SUCCESS]   3 JS files converted to TS
[SUCCESS] 
[SUCCESS]   4 errors caught by rehearsal
[SUCCESS] 
[SUCCESS]   3 have been fixed by rehearsal
[SUCCESS] 
[SUCCESS]   1 errors need to be fixed manually
[SUCCESS] 
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS] 
[SUCCESS]     -- 1 eslint errors, with details in the report
[SUCCESS] 
[STARTED] Regenerating report for TS errors and Eslint errors
[DATA] processing file: /foo.ts
[DATA] processing file: /depends-on-foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Report Generated
[TITLE] 
[TITLE] 
[TITLE]   3 ts files have been scanned
[TITLE] 
[TITLE]   1 errors caught by rehearsal
[TITLE] 
[TITLE]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE] 
[TITLE]     -- 1 eslint errors, with details in the report
[TITLE] 
[SUCCESS] Migration Report Generated
[SUCCESS] 
[SUCCESS] 
[SUCCESS]   3 ts files have been scanned
[SUCCESS] 
[SUCCESS]   1 errors caught by rehearsal
[SUCCESS] 
[SUCCESS]     -- 0 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS] 
[SUCCESS]     -- 1 eslint errors, with details in the report
[SUCCESS] 
"
`;
